File: repos/openrappter/typescript/src/rappids/stats.ts
/**
 * Creature stats, derived — never asserted, never estimated.
 *
 * Weight is exact integer bytes over unique verified content addresses. A
 * `(dimension, sha256)` pair counts once, so carrying the same asset twice
 * cannot make an organism heavier, and an unknown size makes the total
 * *incomplete* rather than approximate: `null` is the honest answer and a
 * guess is not.
 *
 * Lifecycle stage is derived state. It is a reading of how much verified body
 * an organism has accumulated, and it never touches the RAPPID — a Raptor and
 * the baby it grew from are the same creature with the same identity. Nothing
 * here infers maturity from file size.
 *
 * Mirrored by `python/openrappter/rappids/stats.py`.
 */

import { rappCanonicalJson } from './canonical.js';
import { bodyFrameToJson } from './store.js';
import type {
  BodyFrame,
  CreatureStats,
  DimensionSummary,
  LifecycleStage,
  LoadedOrganism,
  QuantumRappidSummary,
  VerificationReport,
} from './types.js';

/**
 * A versioned presentation curve over frame height.
 *
 * Height is how the habitat draws the creature. It is not identity and not a
 * physical fact, so it is versioned and kept away from every integrity path.
 */
export const SPECIES_HEIGHT_CURVE = {
  id: 'quantum-continuity/1',
  baseMm: 420,
  perFrameMm: 90,
  maxMm: 2400,
} as const;

/**
 * Frames that record evidence about the organism rather than a new sense.
 *
 * A census frame proves what verified today; it is not a dimension the
 * creature carries, and counting it as one would inflate the only stat that is
 * supposed to mean "this organism gained a way of being in the world".
 */
export const CENSUS_DIMENSION = 'census';

/**
 * What each stage requires, in the order they are tested.
 *
 * The thresholds are about *verified body*, which is why a compact organism
 * with one rich dimension stays a baby: the point of the ladder is that a
 * Raptor has proven durable memory and recorded skills, not that it has a
 * large file somewhere.
 */
export const STAGE_LADDER: ReadonlyArray<{
  stage: LifecycleStage;
  minFrameHeight: number;
  minActiveDimensions: number;
  requiredActive: readonly string[];
}> = [
  { stage: 'raptor', minFrameHeight: 8, minActiveDimensions: 4, requiredActive: ['memory', 'skill'] },
  { stage: 'hatchling', minFrameHeight: 2, minActiveDimensions: 2, requiredActive: [] },
  { stage: 'baby', minFrameHeight: 0, minActiveDimensions: 0, requiredActive: [] },
];

/**
 * How deep the accepted, unbroken history goes.
 *
 * Contiguous: the first frame that does not continue the chain ends the count,
 * because everything after it rests on a link that did not hold.
 */
export function contiguousFrameHeight(frames: readonly BodyFrame[]): number {
  let height = 0;
  let parent: string | null = null;
  for (const frame of frames) {
    if (frame.seq !== height) break;
    if (frame.prev !== parent) break;
    height += 1;
    parent = frame.payload_hash;
  }
  return height;
}

/** Bytes an accepted frame contributes: its own canonical body, counted once. */
function frameWeightBytes(frames: readonly BodyFrame[], height: number): number {
  const counted = new Set<string>();
  let total = 0;
  for (const frame of frames.slice(0, height)) {
    if (counted.has(frame.frame_hash)) continue;
    counted.add(frame.frame_hash);
    total += Buffer.byteLength(
      rappCanonicalJson(bodyFrameToJson(frame)),
      'utf8',
    );
  }
  return total;
}

export interface DimensionState extends DimensionSummary {
  /** True when this dimension points at content whose size is not known here. */
  unmeasured: boolean;
}

/**
 * Every dimension family the organism carries, declared or witnessed.
 *
 * Declared dimensions come from `rappid.json`; frames can witness families
 * that were not declared when the organism was minted, which is exactly what
 * growth is. Both are the same organism, so both are listed.
 */
export function dimensionStates(
  organism: LoadedOrganism,
  report: VerificationReport,
): DimensionState[] {
  const states = new Map<string, DimensionState>();
  const verified = new Set(
    report.assets
      .filter((asset) => asset.status === 'verified')
      .map((asset) => asset.dimension),
  );
  const brokenDimensions = new Set(
    report.assets.filter((asset) => asset.status !== 'verified').map((asset) => asset.dimension),
  );
  const failedChecks = new Set(
    report.checks.filter((check) => check.status === 'fail').map((check) => check.name.split('.')[0]),
  );

  for (const dimension of organism.document.dimensions) {
    const pathRefs = Object.values(dimension.refs).filter((ref) => ref.includes('/'));
    const externalRefs = Object.values(dimension.refs).filter((ref) => !ref.includes('/'));
    const hasLocalContent = verified.has(dimension.name) || pathRefs.length > 0;
    let status: DimensionSummary['status'];
    if (brokenDimensions.has(dimension.name) || failedChecks.has(dimension.name)) status = 'missing';
    else if (hasLocalContent) status = 'active';
    else status = 'linked';
    states.set(dimension.name, {
      name: dimension.name,
      status,
      ...(dimension.mediaTypes.length > 0 ? { mediaTypes: [...dimension.mediaTypes] } : {}),
      unmeasured: externalRefs.length > 0,
    });
  }

  for (const frame of organism.frames) {
    const dimension = frame.payload.dimension;
    if (states.has(dimension)) continue;
    states.set(dimension, {
      name: dimension,
      status: brokenDimensions.has(dimension) ? 'missing' : 'active',
      unmeasured: false,
    });
  }

  return [...states.values()].sort((left, right) =>
    left.name < right.name ? -1 : left.name > right.name ? 1 : 0,
  );
}

export function deriveStats(
  organism: LoadedOrganism,
  report: VerificationReport,
  dimensions: readonly DimensionState[],
): CreatureStats {
  const frameHeight = contiguousFrameHeight(organism.frames);
  const uniqueVerified = new Map<string, number>();
  for (const asset of report.assets) {
    if (asset.status !== 'verified') continue;
    uniqueVerified.set(
      `${asset.addressSpace}:${asset.addressHash}`,
      asset.expectedBytes,
    );
  }
  let assetWeight = 0;
  for (const bytes of uniqueVerified.values()) assetWeight += bytes;
  const residentWeightBytes = assetWeight + frameWeightBytes(organism.frames, frameHeight);

  // Known-but-absent bytes are only *known* if the manifest that recorded them
  // verified. Otherwise there is no attestation behind the number at all.
  const manifestTrusted = !report.checks.some(
    (check) => check.status === 'fail' && check.name === 'sonic.manifest',
  );
  const linkedUnique = new Map<string, number>();
  if (manifestTrusted) {
    for (const asset of report.assets) {
      if (asset.status !== 'missing') continue;
      const contentKey = asset.addressHash
        ? `${asset.addressSpace}:${asset.addressHash}`
        : `sha256:${asset.expectedSha256}`;
      linkedUnique.set(
        contentKey,
        asset.expectedBytes,
      );
    }
  }
  let linkedKnown = 0;
  for (const bytes of linkedUnique.values()) linkedKnown += bytes;

  const unmeasured = dimensions.some((dimension) => dimension.unmeasured);
  const linkedWeightBytes = unmeasured ? null : linkedKnown;
  const weightComplete = !unmeasured;

  return {
    frameHeight,
    displayHeightMm: Math.min(
      SPECIES_HEIGHT_CURVE.maxMm,
      SPECIES_HEIGHT_CURVE.baseMm + SPECIES_HEIGHT_CURVE.perFrameMm * frameHeight,
    ),
    totalWeightBytes: linkedWeightBytes === null ? null : residentWeightBytes + linkedWeightBytes,
    verifiedWeightBytes: residentWeightBytes,
    residentWeightBytes,
    linkedWeightBytes,
    weightComplete,
    uniqueFrames: new Set(
      organism.frames.slice(0, frameHeight).map((frame) => frame.frame_hash),
    ).size,
    uniqueAssets: uniqueVerified.size,
  };
}

/** Stage is read off verified body. It never re-mints or renames anything. */
export function deriveStageFromEvidence(
  frameHeight: number,
  dimensions: readonly DimensionState[],
): LifecycleStage {
  const active = dimensions.filter((dimension) => dimension.status === 'active');
  const activeNames = new Set(active.map((dimension) => dimension.name));
  for (const rung of STAGE_LADDER) {
    if (frameHeight < rung.minFrameHeight) continue;
    if (active.length < rung.minActiveDimensions) continue;
    if (!rung.requiredActive.every((name) => activeNames.has(name))) continue;
    return rung.stage;
  }
  return 'baby';
}

/** Stage from a complete derived stat block. */
export function deriveStage(
  stats: CreatureStats,
  dimensions: readonly DimensionState[],
): LifecycleStage {
  return deriveStageFromEvidence(stats.frameHeight, dimensions);
}

/** The gateway wire shape for one organism. Identical in both runtimes. */
export function summarize(
  organism: LoadedOrganism,
  report: VerificationReport,
): QuantumRappidSummary {
  const dimensions = dimensionStates(organism, report);
  const stats = deriveStats(organism, report, dimensions);
  const sonic = organism.sonic;
  const verifiedPaths = new Set(
    report.assets.filter((asset) => asset.status === 'verified').map((asset) => asset.path),
  );

  const summary: QuantumRappidSummary = {
    rappid: organism.document.rappid,
    name: organism.document.name,
    displayName: organism.document.displayName,
    species: organism.document.name,
    lifecycleStage: deriveStage(stats, dimensions),
    localOnly: organism.document.localOnly,
    parentRappid: organism.document.parentRappid,
    stats,
    traits: { ...organism.traits.traits },
    dimensions: dimensions.map((dimension) => {
      const entry: DimensionSummary = { name: dimension.name, status: dimension.status };
      if (dimension.mediaTypes !== undefined) entry.mediaTypes = [...dimension.mediaTypes];
      return entry;
    }),
    externalEpisode: organism.document.externalEpisode,
    verified: report.ok,
    unmeasuredDimensions: dimensions
      .filter((dimension) => dimension.unmeasured)
      .map((dimension) => dimension.name),
  };

  if (sonic !== null) {
    const playback = sonic.devicePlayback;
    const verifiedTrack = (path: string | null): boolean =>
      path !== null && verifiedPaths.has(path);
    summary.sonic = {
      wakeCall: verifiedTrack(playback.preferred) || verifiedTrack(playback.losslessFallback),
      midiDna: verifiedTrack(playback.midiPrompt),
      autocomplete: verifiedTrack(playback.midiAutocomplete),
    };
  }
  return summary;
}

File: repos/openrappter/python/openrappter/rappids/stats.py
"""Creature stats, derived -- never asserted, never estimated.

Mirrors ``typescript/src/rappids/stats.ts``.

Weight is exact integer bytes over unique verified content addresses. A
``(dimension, sha256)`` pair counts once, so carrying the same asset twice
cannot make an organism heavier, and an unknown size makes the total
*incomplete* rather than approximate: ``None`` is the honest answer and a guess
is not.

Lifecycle stage is derived state. It is a reading of how much verified body an
organism has accumulated, and it never touches the RAPPID -- a Raptor and the
baby it grew from are the same creature with the same identity. Nothing here
infers maturity from file size.
"""

from __future__ import annotations

from typing import Any, Dict, List, Sequence

from .canonical import rapp_canonical_json
from .store import body_frame_to_json
from .types import (
    BodyFrame,
    CreatureStats,
    DimensionState,
    LoadedOrganism,
    VerificationReport,
)

#: A versioned presentation curve over frame height. Height is how the habitat
#: draws the creature: not identity, not a physical fact, and kept away from
#: every integrity path.
SPECIES_HEIGHT_CURVE = {
    "id": "quantum-continuity/1",
    "base_mm": 420,
    "per_frame_mm": 90,
    "max_mm": 2400,
}

#: Frames that record evidence about the organism rather than a new sense. A
#: census frame proves what verified today; counting it as a dimension would
#: inflate the one stat that is supposed to mean "this organism gained a way of
#: being in the world".
CENSUS_DIMENSION = "census"

#: What each stage requires, in the order they are tested. The thresholds are
#: about *verified body*, which is why a compact organism with one rich
#: dimension stays a baby.
STAGE_LADDER = (
    {
        "stage": "raptor",
        "min_frame_height": 8,
        "min_active_dimensions": 4,
        "required_active": ("memory", "skill"),
    },
    {
        "stage": "hatchling",
        "min_frame_height": 2,
        "min_active_dimensions": 2,
        "required_active": (),
    },
    {
        "stage": "baby",
        "min_frame_height": 0,
        "min_active_dimensions": 0,
        "required_active": (),
    },
)


def contiguous_frame_height(frames: Sequence[BodyFrame]) -> int:
    """How deep the accepted, unbroken history goes.

    Contiguous: the first frame that does not continue the chain ends the
    count, because everything after it rests on a link that did not hold.
    """
    height = 0
    parent = None
    for frame in frames:
        if frame.seq != height:
            break
        if frame.prev != parent:
            break
        height += 1
        parent = frame.payload_hash
    return height


def _frame_weight_bytes(frames: Sequence[BodyFrame], height: int) -> int:
    """Bytes an accepted frame contributes: its own canonical body, counted once."""
    counted = set()
    total = 0
    for frame in frames[:height]:
        if frame.frame_hash in counted:
            continue
        counted.add(frame.frame_hash)
        total += len(rapp_canonical_json(body_frame_to_json(frame)).encode("utf-8"))
    return total


def dimension_states(
    organism: LoadedOrganism, report: VerificationReport
) -> List[DimensionState]:
    """Every dimension family the organism carries, declared or witnessed.

    Declared dimensions come from ``rappid.json``; frames can witness families
    that were not declared when the organism was minted, which is exactly what
    growth is. Both are the same organism, so both are listed.
    """
    states: Dict[str, DimensionState] = {}
    verified = {asset.dimension for asset in report.assets if asset.status == "verified"}
    broken = {asset.dimension for asset in report.assets if asset.status != "verified"}
    failed = {
        check.name.split(".", 1)[0] for check in report.checks if check.status == "fail"
    }

    for dimension in organism.document.dimensions:
        path_refs = [ref for ref in dimension.refs.values() if "/" in ref]
        external_refs = [ref for ref in dimension.refs.values() if "/" not in ref]
        has_local_content = dimension.name in verified or bool(path_refs)
        if dimension.name in broken or dimension.name in failed:
            status = "missing"
        elif has_local_content:
            status = "active"
        else:
            status = "linked"
        states[dimension.name] = DimensionState(
            name=dimension.name,
            status=status,
            media_types=list(dimension.media_types),
            unmeasured=bool(external_refs),
        )

    for frame in organism.frames:
        dimension_name = str(frame.payload.get("dimension"))
        if dimension_name in states:
            continue
        states[dimension_name] = DimensionState(
            name=dimension_name,
            status="missing" if dimension_name in broken else "active",
            media_types=[],
            unmeasured=False,
        )

    return [states[name] for name in sorted(states)]


def derive_stats(
    organism: LoadedOrganism,
    report: VerificationReport,
    dimensions: Sequence[DimensionState],
) -> CreatureStats:
    frame_height = contiguous_frame_height(organism.frames)
    unique_verified: Dict[str, int] = {}
    for asset in report.assets:
        if asset.status != "verified":
            continue
        unique_verified[f"{asset.address_space}:{asset.address_hash}"] = asset.expected_bytes
    resident_weight_bytes = sum(unique_verified.values()) + _frame_weight_bytes(
        organism.frames, frame_height
    )

    # Known-but-absent bytes are only *known* if the manifest that recorded
    # them verified. Otherwise there is no attestation behind the number.
    manifest_trusted = not any(
        check.status == "fail" and check.name == "sonic.manifest" for check in report.checks
    )
    linked_unique: Dict[str, int] = {}
    if manifest_trusted:
        for asset in report.assets:
            if asset.status != "missing":
                continue
            # A missing asset has no egg address to count by -- nothing hashed
            # it, because the bytes are not here. Its manifest digest is the
            # only identity it has, and it still must count exactly once.
            content_key = (
                f"{asset.address_space}:{asset.address_hash}"
                if asset.address_hash
                else f"sha256:{asset.expected_sha256}"
            )
            linked_unique[content_key] = asset.expected_bytes
    linked_known = sum(linked_unique.values())

    unmeasured = any(dimension.unmeasured for dimension in dimensions)
    linked_weight_bytes = None if unmeasured else linked_known

    return CreatureStats(
        frame_height=frame_height,
        display_height_mm=min(
            int(SPECIES_HEIGHT_CURVE["max_mm"]),
            int(SPECIES_HEIGHT_CURVE["base_mm"])
            + int(SPECIES_HEIGHT_CURVE["per_frame_mm"]) * frame_height,
        ),
        total_weight_bytes=(
            None if linked_weight_bytes is None else resident_weight_bytes + linked_weight_bytes
        ),
        verified_weight_bytes=resident_weight_bytes,
        resident_weight_bytes=resident_weight_bytes,
        linked_weight_bytes=linked_weight_bytes,
        weight_complete=not unmeasured,
        unique_frames=len({frame.frame_hash for frame in organism.frames[:frame_height]}),
        unique_assets=len(unique_verified),
    )


def derive_stage_from_evidence(
    frame_height: int, dimensions: Sequence[DimensionState]
) -> str:
    """Stage is read off verified body. It never re-mints or renames anything."""
    active = [dimension for dimension in dimensions if dimension.status == "active"]
    active_names = {dimension.name for dimension in active}
    for rung in STAGE_LADDER:
        if frame_height < rung["min_frame_height"]:
            continue
        if len(active) < rung["min_active_dimensions"]:
            continue
        if not all(name in active_names for name in rung["required_active"]):
            continue
        return str(rung["stage"])
    return "baby"


def derive_stage(stats: CreatureStats, dimensions: Sequence[DimensionState]) -> str:
    """Stage from a complete derived stat block."""
    return derive_stage_from_evidence(stats.frame_height, dimensions)


def summarize(organism: LoadedOrganism, report: VerificationReport) -> Dict[str, Any]:
    """The gateway wire shape for one organism. Identical in both runtimes."""
    dimensions = dimension_states(organism, report)
    stats = derive_stats(organism, report, dimensions)
    sonic = organism.sonic
    verified_paths = {
        asset.path for asset in report.assets if asset.status == "verified"
    }

    summary: Dict[str, Any] = {
        "rappid": organism.document.rappid,
        "name": organism.document.name,
        "displayName": organism.document.display_name,
        "species": organism.document.name,
        "lifecycleStage": derive_stage(stats, dimensions),
        "localOnly": organism.document.local_only,
        "parentRappid": organism.document.parent_rappid,
        "stats": stats.to_wire(),
        "traits": dict(organism.traits.traits),
        "dimensions": [dimension.to_wire() for dimension in dimensions],
        "externalEpisode": (
            None
            if organism.document.external_episode is None
            else organism.document.external_episode.to_wire()
        ),
        "verified": report.ok,
        "unmeasuredDimensions": [
            dimension.name for dimension in dimensions if dimension.unmeasured
        ],
    }

    if sonic is not None:
        playback = sonic.device_playback

        def verified_track(path):
            return path is not None and path in verified_paths

        summary["sonic"] = {
            "wakeCall": verified_track(playback.preferred)
            or verified_track(playback.lossless_fallback),
            "midiDna": verified_track(playback.midi_prompt),
            "autocomplete": verified_track(playback.midi_autocomplete),
        }
    return summary
